sw/device/lib/dif/dif_rstmgr.c

```diff
@@ -10,84 +10,86 @@
-    .mask = RSTMGR_RESET_INFO_CAUSE_MASK,
-    .index = RSTMGR_RESET_INFO_CAUSE_OFFSET,
-};
-
-// Regen registers are "write 1 to clear" = write 1 to lock.
-static const ptrdiff_t kRstmgrRegenOffsetsMap[] = {
-        [kDifRstmgrPeripheralSpi] = RSTMGR_SPI_DEVICE_REGEN_REG_OFFSET,
-        [kDifRstmgrPeripheralUsb] = RSTMGR_USB_REGEN_REG_OFFSET,
-};
-
-// Reset registers, when set to 0, peripheral is held in reset.
-static const ptrdiff_t kRstmgrResetOffsetsMap[] = {
-        [kDifRstmgrPeripheralSpi] = RSTMGR_RST_SPI_DEVICE_N_REG_OFFSET,
-        [kDifRstmgrPeripheralUsb] = RSTMGR_RST_USB_N_REG_OFFSET,
-};
-
-// TODO - static assert to make sure that the above arrays cover all relevant
-//        peripherals?
+    RSTMGR_PARAM_NUMSWRESETS <= 32,
+    "Reset Enable and Control registers span across multiple registers!");
+
+/**
+ * Get Reset Enable and Reset Control register bit index.
+ *
+ * This function calculates the bit position within a register for a specific
+ * peripheral.
+ */
+static bitfield_bit32_index_t rstmgr_regen_rstctrl_bit_index(
+    dif_rstmgr_peripheral_t id) {
+  return id % RSTMGR_PARAM_REG_WIDTH;
+}
+
+static bool rstmgr_software_reset_is_locked(
+    mmio_region_t base_addr, dif_rstmgr_peripheral_t peripheral) {
+  uint32_t bitfield =
+      mmio_region_read32(base_addr, RSTMGR_SW_RST_REGEN_REG_OFFSET);
+  bitfield_bit32_index_t index = rstmgr_regen_rstctrl_bit_index(peripheral);
+
+  // When bit is cleared, the software reset for `peripheral` is disabled.
+  return !bitfield_bit32_read(bitfield, index);
+}
 
 static void rstmgr_software_reset_hold(mmio_region_t base_addr,
                                        dif_rstmgr_peripheral_t peripheral) {
-  ptrdiff_t reg_offset = kRstmgrResetOffsetsMap[peripheral];
-  uint32_t bitfield = bitfield_bit32_write(0, RSTMGR_RESET_BIT_OFFSET, false);
-  mmio_region_write32(base_addr, reg_offset, bitfield);
+  uint32_t bitfield =
+      mmio_region_read32(base_addr, RSTMGR_SW_RST_CTRL_N_REG_OFFSET);
+
+  bitfield_bit32_index_t index = rstmgr_regen_rstctrl_bit_index(peripheral);
+  bitfield = bitfield_bit32_write(bitfield, index, false);
+
+  // Clear bit to hold `peripheral` in the reset state.
+  mmio_region_write32(base_addr, RSTMGR_SW_RST_CTRL_N_REG_OFFSET, bitfield);
 }
 
-static bool rstmgr_software_reset_is_held(mmio_region_t base_addr,
-                                          dif_rstmgr_peripheral_t peripheral) {
-  ptrdiff_t reg_offset = kRstmgrResetOffsetsMap[peripheral];
-  uint32_t bitfield = mmio_region_read32(base_addr, reg_offset);
-  return !bitfield_bit32_read(bitfield, RSTMGR_RESET_BIT_OFFSET);
+/**
+ * Clears entire reset info register.
+ *
+ * Normal "Power On Reset" cause is also cleared. Set bit to clear.
+ */
+void rstmgr_reset_info_clear(mmio_region_t base_addr) {
+  mmio_region_write32(base_addr, RSTMGR_RESET_INFO_REG_OFFSET, UINT32_MAX);
 }
 
 static void rstmgr_software_reset_release(mmio_region_t base_addr,
```

Comment:
nit: You can remove one of `_reset_hold` or `_reset_release` by adding a new parameter for the new state.

Reply:
Addes `bool hold` to the `hold` function, which then determines to hold or release the reset.